### PR TITLE
fix cert dir in kubeadm

### LINKED ISSUE #232: it should be trivial to se set alternative config path (not /etc/kubernetes)
I'm using `kubeadm` on system where `/etc` is mounted read-only, and it was non-trivial to set alternative config path.

Namely, I had to do this:
```
env KUBE_KUBERNETES_DIR=/var/lib/kubeadm kubeadm init --skip-preflight-checks --kubernetes-version v1.6.1 --cert-dir /var/lib/kubeadm/pki
```
It should be just one argument. The main question is – why do we have separate `--cert-dir` flag?

There other good reasons why someone may wish to configure alternative path for configuration.

### PR DESCRIPTION
**Pull request by xilabao**
1.fixes https://github.com/kubernetes/kubeadm/issues/232
2. use manifests as a constant

### CODE REVIEW

**Comment by spiffxp** [comment]
@k8s-bot ok to test

**Comment by xilabao** [comment]
@dmmcquay @mikedanese kindly ping

**Comment by fabriziopandini** [comment]
@xilabao, could you kindly confirm if this PR fixes the behaviour of kubeadm flags currently insensible to KUBE_KUBERNETES_DIR env var, and more specifically:
- CertificatesDir which always defaults to /etc/kubernetes/pk,
- CACertPath  which always defaults to "/etc/kubernetes/pki/ca.crt"

If yes this would be great, because this was one of the planned activities for [Kubeadm phase implementation checklist](https://github.com/kubernetes/kubeadm/issues/267)

File: cmd/kubeadm/app/cmd/phases/kubeconfig.go [hunk 1]
```diff
@@ -74,7 +73,7 @@ func NewCmdClientCerts(out io.Writer) *cobra.Command {
 }
 
 func addCommonFlags(cmd *cobra.Command, config *kubeconfigphase.BuildConfigProperties) {
-	cmd.Flags().StringVar(&config.CertDir, "cert-dir", kubeadmapiext.DefaultCertificatesDir, "The path to the directory where the certificates are.")
+	cmd.Flags().StringVar(&config.CertDir, "cert-dir", "", "The path to the directory where the certificates are.")
```
Comment (mikedanese) [inline cmd/kubeadm/app/cmd/phases/kubeconfig.go]: Why default to the current directory?

Comment (xilabao) [inline cmd/kubeadm/app/cmd/phases/kubeconfig.go]: Restored. PTAL

**Comment by xilabao** [comment]
@fabriziopandini I think this PR has fixed it.

File: cmd/kubeadm/app/cmd/init.go [hunk 2]
```diff
@@ -157,6 +157,10 @@ func NewInit(cfgPath string, cfg *kubeadmapi.MasterConfiguration, skipPreFlight,
 		}
 	}
 
+	if len(cfg.CertificatesDir) == 0 {
```
Comment (dmmcquay) [inline cmd/kubeadm/app/cmd/init.go]: Shouldn't this happen in apis/kubeadm/v1alpha1/defaults.go?

File: cmd/kubeadm/app/cmd/join.go [hunk 3]
```diff
@@ -144,6 +144,10 @@ func NewJoin(cfgPath string, args []string, cfg *kubeadmapi.NodeConfiguration, s
 		}
 	}
 
+	if len(cfg.CACertPath) == 0 {
```
Comment (dmmcquay) [inline cmd/kubeadm/app/cmd/join.go]: same

**Review by dmmcquay — CHANGES_REQUESTED** [review]
Can you be more descriptive in you PR comments of what your change does?

File: cmd/kubeadm/app/cmd/init.go (hunk 2 above)
Comment (xilabao) [inline cmd/kubeadm/app/cmd/init.go]: Done. PTAL

**Comment by xilabao** [comment]
@k8s-bot pull-kubernetes-federation-e2e-gce test this

**Comment by xilabao** [comment]
@mikedanese @dmmcquay kindly ping

File: cmd/kubeadm/app/apis/kubeadm/v1alpha1/defaults.go [hunk 4]
```diff
@@ -18,9 +18,12 @@ package v1alpha1
 
 import (
 	"net/url"
+	"path"
```
Comment (luxas) [inline cmd/kubeadm/app/apis/kubeadm/v1alpha1/defaults.go]: use filepath instead

File: cmd/kubeadm/app/apis/kubeadm/v1alpha1/defaults.go [hunk 5]
```diff
@@ -61,7 +62,7 @@ func SetDefaults_MasterConfiguration(obj *MasterConfiguration) {
 	}
 
 	if obj.CertificatesDir == "" {
-		obj.CertificatesDir = DefaultCertificatesDir
+		obj.CertificatesDir = path.Join(kubeadmapi.GlobalEnvParams.KubernetesDir, kubeadmconstants.CertificatesDirName)
```
Comment (luxas) [inline cmd/kubeadm/app/apis/kubeadm/v1alpha1/defaults.go]: We strive to phase out kubeadmapi.GlobalEnvParams.KubernetesDir and try to use it as little as possible, please revert this

File: cmd/kubeadm/app/apis/kubeadm/v1alpha1/defaults.go [hunk 6]
```diff
@@ -75,8 +76,9 @@ func SetDefaults_MasterConfiguration(obj *MasterConfiguration) {
 
 func SetDefaults_NodeConfiguration(obj *NodeConfiguration) {
 	if obj.CACertPath == "" {
-		obj.CACertPath = DefaultCACertPath
+		obj.CACertPath = path.Join(kubeadmapi.GlobalEnvParams.KubernetesDir, kubeadmconstants.CertificatesDirName, kubeadmconstants.CACertName)
```
Comment (luxas) [inline cmd/kubeadm/app/apis/kubeadm/v1alpha1/defaults.go]: dito

File: cmd/kubeadm/app/cmd/phases/kubeconfig.go [hunk 7]
```diff
@@ -74,7 +76,8 @@ func NewCmdClientCerts(out io.Writer) *cobra.Command {
 }
 
 func addCommonFlags(cmd *cobra.Command, config *kubeconfigphase.BuildConfigProperties) {
-	cmd.Flags().StringVar(&config.CertDir, "cert-dir", kubeadmapiext.DefaultCertificatesDir, "The path to the directory where the certificates are.")
+	certificatesDir := filepath.Join(kubeadmapi.GlobalEnvParams.KubernetesDir, kubeadmconstants.CertificatesDirName)
```
Comment (luxas) [inline cmd/kubeadm/app/cmd/phases/kubeconfig.go]: One important thing to realize is that KubernetesDir and CertificatesDir should be able to be in two different roots, they should be able to be decoupled.
i.e. KubernetesDir could be in /etc/kubernetes (the default) but CertificatesDir /var/lib/kube-certificates

File: cmd/kubeadm/app/constants/constants.go [hunk 8]
```diff
@@ -25,7 +25,9 @@ import (
 
 const (
 	// KubernetesDir is the directory kubernetes owns for storing various configuration files
-	KubernetesDir = "/etc/kubernetes"
+	KubernetesDir       = "/etc/kubernetes"
+	CertificatesDirName = "pki"
```
Comment (luxas) [inline cmd/kubeadm/app/constants/constants.go]: See comment above, these are not coupled to each other.

**Review by luxas — CHANGES_REQUESTED** [review]
One important thing to realize is that KubernetesDir and CertificatesDir should be able to be in two different roots, they should be able to be decoupled.
i.e. KubernetesDir could be in /etc/kubernetes (the default) but CertificatesDir /var/lib/kube-certificates

Please revert the changes touching /etc/kubernetes/pki.
I'm open to making "manifests" a constant though

**Comment by luxas** [comment]
also needs a rebase

**Comment by xilabao** [comment]
@luxas @dmmcquay PTAL

File: cmd/kubeadm/app/constants/constants.go [hunk 9]
```diff
@@ -27,6 +27,8 @@ const (
 	// KubernetesDir is the directory kubernetes owns for storing various configuration files
 	KubernetesDir = "/etc/kubernetes"
 
+	ManifestsDirName = "manifests"
```
Comment (luxas) [inline cmd/kubeadm/app/constants/constants.go]: nit: I think I'd prefer `ManifestsSubDirName`, sorry for not catching that the first pass

**Review by luxas — COMMENTED** [review]
LGTM but one nit

File: cmd/kubeadm/app/constants/constants.go (hunk 9 above)
Comment (xilabao) [inline cmd/kubeadm/app/constants/constants.go]: Done. PTAL

**Review by luxas — APPROVED** [review]
LGTM

**Comment by luxas** [comment]
/lgtm

@k8s-bot pull-kubernetes-federation-e2e-gce test this

**Comment by luxas** [comment]
@k8s-bot pull-kubernetes-federation-e2e-gce test this

**Pull request merged**